smdebug/tensorflow/base_hook.py

```diff
@@ -489,12 +491,14 @@ def set_optimizer_variables(self, optimizer_variables):
         # From TF 2.2, executing_eagerly_outside_functions() can be used as
         # ops.executing_eagerly_outside_functions() or tf.compat.v1.executing_eagerly_outside_functions().
         # But in TF 2.1, only ops.executing_eagerly_outside_functions() is valid
-        if is_tf_version_2x() and ops.executing_eagerly_outside_functions():
-            return
         # since this is done for each variable at a time for keras, not checking if set already
         self.collection_manager.get(CollectionKeys.OPTIMIZER_VARIABLES).add_for_mode(
             optimizer_variables, ModeKeys.TRAIN
         )
+        if self.save_all is True:
```

Comment:
earlier save_all method was not saving optimizer variables?

Reply:
In Eager mode optimizer variables only become available in `_on_any_batch_end`. 
The population tensor refs to the ALL collection was happening only once in `_on_any_batch_begin`.

But since the ALL collection is not the only additional collection that might need to be populated with optimizer variables, I have added an additional `_prepare_non_layer_tensors()` call in `_on_any_batch_end`